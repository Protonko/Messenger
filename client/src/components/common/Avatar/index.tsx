import {useMemo, FC} from 'react'
import classNames from 'classnames'
import {Sizes} from 'models/common/sizes'
import {ColorGeneratorByName} from 'utils/ColorGeneratorByName'
import {isDarkColor} from 'utils/isDarkColor'
import {Loader} from 'components/common/Loader'

export interface IPropsAvatar {
  src: string
  name?: string
  size?: Sizes
  customStyles?: string
}

export const Avatar: FC<IPropsAvatar> = ({
  src,
  name,
  size = Sizes.MEDIUM,
  customStyles,
}) => {
  const backgroundColor = useMemo(
    () => new ColorGeneratorByName(name ?? '').generate(),
    [src, name],
  )
  const avatarClassname = classNames(
    'avatar',
    {'avatar--sm': size === Sizes.SMALL},
    {'avatar--lg': size === Sizes.LARGE},
    {[customStyles ?? '']: customStyles},
  )

  if (!src && name) {
    const classNamesText = classNames(
      'avatar__text',
      {'avatar__text--light': isDarkColor(backgroundColor)},
      {'avatar__text--dark': !isDarkColor(backgroundColor)},
    )

    return (
      <div className={avatarClassname} style={{backgroundColor}}>
        <span className={classNamesText}>{name.charAt(0)}</span>
      </div>
    )
  }

  return (
    <div className={avatarClassname}>
      <img className="avatar__image" src={src} alt="" />
      <Loader />
    </div>
  )
}
