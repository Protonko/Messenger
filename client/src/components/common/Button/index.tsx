import {FC, HTMLProps, ReactNode} from 'react'

interface IButtonProps extends HTMLProps<HTMLButtonElement> {
  text?: string
  icon?: ReactNode
  leftAddons?: ReactNode
  rightAddons?: ReactNode
  type?: 'button' | 'submit' | 'reset'
}

export const Button: FC<IButtonProps> = ({
  text,
  icon,
  leftAddons,
  rightAddons,
  ...buttonProps
}) => {
  const renderButtonContent = () => {
    return (
      <>
        {text && (
          <span className="button__text">
            {text}
          </span>
        )}
        {icon && (
          <span className="button__icon">
            {icon}
          </span>
        )}
      </>
    )
  }

  return (
    <button className="button" {...buttonProps}>
      {leftAddons}
      {renderButtonContent()}
      {rightAddons}
    </button>
  )
}
