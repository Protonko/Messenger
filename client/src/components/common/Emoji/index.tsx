import 'emoji-mart/css/emoji-mart.css'
import {FC, useState, useRef} from 'react'
import {EmojiData, Picker} from 'emoji-mart'
import {ReactComponent as Smile} from 'assets/icons/smile.svg'
import {useOutsideClick} from 'hooks/useOutsideClick'

export interface IPropsEmoji {
  onSelect: (emoji: EmojiData) => void
}

export const Emoji: FC<IPropsEmoji> = ({onSelect}) => {
  const [open, setOpen] = useState(false)
  const ref = useRef(null)
  const initiatorRef = useRef(null)

  useOutsideClick(ref, () => setOpen(false), initiatorRef)

  const onSelectEmoji = (emoji: EmojiData) => {
    onSelect(emoji)
    setOpen(false)
  }

  return (
    <div className="emoji">
      {open && (
        <div className="emoji__popup" ref={ref}>
          <Picker native={true} onSelect={onSelectEmoji} />
        </div>
      )}

      <Smile
        className="emoji__button"
        ref={initiatorRef}
        onClick={() => setOpen(!open)}
      />
    </div>
  )
}
