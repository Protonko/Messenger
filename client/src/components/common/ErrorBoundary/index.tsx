import {Component, ErrorInfo} from 'react';

interface IProps {}
interface IState {
  error: null | Error;
  errorInfo: null | ErrorInfo;
  hasError: boolean;
}

export class ErrorBoundary extends Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);

    this.state = {
      error: null,
      errorInfo: null,
      hasError: false,
    }
  }

  static getDerivedStateFromError(error: Error) {
    console.log(error)
    return {
      hasError: true,
    };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    this.setState({
      error: error,
      errorInfo: errorInfo
    })
  }

  _renderDetailInfo() {
    if (process.env.NODE_ENV === 'development') {
      return (
        <div>

        </div>
      )
    }

    return null
  }

  render() {
    if (this.state.hasError) {
      return (
        <>
          <h1>Что-то пошло не так.</h1>

          {process.env.NODE_ENV === 'development' && (
            <details style={{ whiteSpace: 'pre-wrap' }}>
              {this.state.error && this.state.error.toString()}
              <br />
              {this.state.errorInfo?.componentStack}
            </details>
          )}
        </>
      )
    }

    return this.props.children;
  }
}
