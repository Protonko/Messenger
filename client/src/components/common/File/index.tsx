import {useEffect, useState, FC} from 'react'
import classNames from 'classnames'
import {ReactComponent as FileIcon} from 'assets/icons/file.svg'

export interface IFileProps {
  file: File
  link?: string
  onRemove?: (name: string, lastModified: number) => void
  additionalClassName?: string
}

export const File: FC<IFileProps> = ({file, onRemove, additionalClassName}) => {
  const [preview, setPreview] = useState('')
  const isImage = file.type.startsWith('image')
  const fileClassNames = classNames(
    'file-preview',
    {'file-preview--image': isImage},
    {'file-preview--file': !isImage},
    {[additionalClassName ?? '']: !!additionalClassName},
  )

  const convertToMB = (bytes: number) => {
    const CONVERSATION_VALUE = 1024
    return (bytes / CONVERSATION_VALUE ** 2).toFixed(2)
  }

  useEffect(() => {
    const objectUrl = URL.createObjectURL(new Blob([file]))
    setPreview(objectUrl)

    return () => URL.revokeObjectURL(objectUrl)
  }, [file])

  const renderRemoveIcon = () => {
    if (onRemove) {
      return (
        <div
          className="file-preview__remove"
          onClick={() => onRemove(file.name, file.lastModified)}
        >
          &times;
        </div>
      )
    }

    return null
  }

  if (file.type.startsWith('image')) {
    return (
      <div
        className={fileClassNames}
        style={{backgroundImage: `url(${preview})`}}
      >
        {renderRemoveIcon()}
      </div>
    )
  }

  return (
    <div className={fileClassNames}>
      {renderRemoveIcon()}

      <FileIcon className="file-preview__icon" />
      <div className="file-preview__info">
        <span className="file-preview__info-name">{file.name}</span>
        <span className="file-preview__info-size">
          {convertToMB(file.size)} MB
        </span>
      </div>
    </div>
  )
}
