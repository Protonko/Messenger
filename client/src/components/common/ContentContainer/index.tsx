import type {FC} from 'react'
import {Loader} from 'components/common/Loader'
import {Text} from 'components/common/Text'

interface ContentContainerProps {
  loading: boolean
  errorMessage: string | null
  children: JSX.Element
}

export const ContentContainer: FC<ContentContainerProps> = ({
  loading,
  errorMessage,
  children,
}) => {
  if (errorMessage) {
    return <Text>{errorMessage}</Text>
  }

  if (loading) {
    return <Loader center={true} />
  }

  return children
}
