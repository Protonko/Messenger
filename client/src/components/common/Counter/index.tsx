// types
import {Status} from 'models/common/status'

import {FC} from 'react'
import classNames from 'classnames'

export interface IPropsCounter {
  count: number | '99+',
  status?: Status,
}

export const Counter: FC<IPropsCounter> = ({
  count,
  status = Status.ACTIVE,
}) => {
  const classNameStatus = classNames(
    'counter',
    {'counter--muted': status === Status.MUTED}
  )

  return (
    <div className={classNameStatus}>
      <span className="counter__text">
        {count}
      </span>
    </div>
  )
}
