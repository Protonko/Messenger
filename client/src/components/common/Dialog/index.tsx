import type {FC} from 'react'
import type {IDialog} from 'models/dialog'
import {ReadStatus} from 'models/common/status'
import {TextSize, TextTypes} from 'models/common/text'
import {ReactComponent as DoubleCheck} from 'assets/icons/double-check.svg'
import {ReactComponent as Check} from 'assets/icons/check.svg'
import {COLORS} from 'static/colors'
import {Text} from 'components/common/Text'
import {Counter} from 'components/common/Counter'
import {Avatar} from 'components/common/Avatar'

export const Dialog: FC<IDialog> = ({
  name,
  description,
  avatar,
  edited,
  createdAt,
  updatedAt,
  messages,
  status,
  readStatus,
}) => {
  const counter = messages.toString().length > 2 ? '99+' : messages

  const renderStatus = () => {
    if (!!counter) {
      return <Counter count={counter} status={status}/>
    }

    if (readStatus) {
      const props = {
        width: 15,
        height: 15,
        color: COLORS.dustyGray,
      }
      return readStatus === ReadStatus.READ ? <DoubleCheck {...props} /> : <Check {...props} />
    }

    return null
  }

  return (
    <div className="dialog">
      <Avatar src={avatar} />

      <div className="dialog__data">
        <div className="dialog__text">
          <Text
            type={TextTypes.h4}
            customStyles="dialog__text-title"
            numberOfLines={1}
          >
            {name}
          </Text>
          <Text>{description}</Text>
        </div>

        <div className="dialog__info">
          <Text
            type={TextTypes.mixed}
            size={TextSize.EXTRA_SMALL}
            customStyles="dialog__info-date"
          >
            {new Date(edited ? updatedAt : createdAt).toLocaleDateString()}
          </Text>

          {renderStatus()}
        </div>
      </div>
    </div>
  )
}
