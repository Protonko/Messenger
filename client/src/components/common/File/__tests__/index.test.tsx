import {shallow, ShallowWrapper} from 'enzyme'
import {File as FileComponent, IFileProps} from 'components/common/File'

const ELEMENT_SELECTORS = {
  file: '.file-preview',
  removeButton: '.file-preview__remove',
  imageFile: '.file-preview--image',
  additional: 'additional-classname',
}

describe('File', () => {
  let component: ShallowWrapper<IFileProps>
  let props: IFileProps

  beforeEach(() => {
    props = {
      file: new File([], 'mock'),
      onRemove: jest.fn(),
    }
  })

  it('Should render File component', () => {
    component = shallow(<FileComponent {...props} />)

    expect(component).toMatchSnapshot('File')
    expect(component.find(ELEMENT_SELECTORS.file).length).toBe(1)
  })

  it('Should render File component with additional classname', () => {
    component = shallow(
      <FileComponent
        {...props}
        additionalClassName={ELEMENT_SELECTORS.additional}
      />,
    )

    expect(component).toMatchSnapshot('File with additional classname')
    expect(component.find(`.${ELEMENT_SELECTORS.additional}`).length).toBe(1)
  })

  it('Should render File without remove button', () => {
    component = shallow(<FileComponent {...props} onRemove={undefined} />)

    expect(component).toMatchSnapshot('File without remove button')
    expect(component.find(ELEMENT_SELECTORS.removeButton).length).toBe(0)
  })

  it('Should render image File', () => {
    const image = new File([], 'image.png', {type: 'image/png'})
    component = shallow(<FileComponent {...props} file={image} />)

    expect(component).toMatchSnapshot('File with image file')
    expect(component.find(ELEMENT_SELECTORS.imageFile).length).toBe(1)
  })

  it('Should call onRemove', () => {
    component = shallow(<FileComponent {...props} />)
    component.find(ELEMENT_SELECTORS.removeButton).simulate('click')

    expect(props.onRemove).toBeCalled()
  })
})
