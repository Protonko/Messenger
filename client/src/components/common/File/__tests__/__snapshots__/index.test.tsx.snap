// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File Should render File component with additional classname: File with additional classname 1`] = `
<div
  className="file-preview file-preview--file additional-classname"
>
  <div
    className="file-preview__remove"
    onClick={[Function]}
  >
    ×
  </div>
  <div
    className="file-preview__icon"
  />
  <div
    className="file-preview__info"
  >
    <span
      className="file-preview__info-name"
    >
      mock
    </span>
    <span
      className="file-preview__info-size"
    >
      0.00
       MB
    </span>
  </div>
</div>
`;

exports[`File Should render File component: File 1`] = `
<div
  className="file-preview file-preview--file"
>
  <div
    className="file-preview__remove"
    onClick={[Function]}
  >
    ×
  </div>
  <div
    className="file-preview__icon"
  />
  <div
    className="file-preview__info"
  >
    <span
      className="file-preview__info-name"
    >
      mock
    </span>
    <span
      className="file-preview__info-size"
    >
      0.00
       MB
    </span>
  </div>
</div>
`;

exports[`File Should render File without remove button: File without remove button 1`] = `
<div
  className="file-preview file-preview--file"
>
  <div
    className="file-preview__icon"
  />
  <div
    className="file-preview__info"
  >
    <span
      className="file-preview__info-name"
    >
      mock
    </span>
    <span
      className="file-preview__info-size"
    >
      0.00
       MB
    </span>
  </div>
</div>
`;

exports[`File Should render image File: File with image file 1`] = `
<div
  className="file-preview file-preview--image"
  style={
    Object {
      "backgroundImage": "url()",
    }
  }
>
  <div
    className="file-preview__remove"
    onClick={[Function]}
  >
    ×
  </div>
</div>
`;
